src/libs/relay/conduit_relay_mpi.cpp

```diff
@@ -113,171 +113,618 @@ rank(MPI_Comm mpi_comm)
+        #endif
+
+    }
+
+    return res;
+}
+
 
 //---------------------------------------------------------------------------//
 int 
-send(Node &node, int dest, int tag, MPI_Comm comm)
+send_using_schema(const Node &node, int dest, int tag, MPI_Comm comm)
 { 
 
-    Schema schema_c;
-    node.schema().compact_to(schema_c);
-    std::string schema = schema_c.to_json();
-    int schema_len = schema.length() + 1;
-
-    std::vector<uint8> data;
-    node.serialize(data);
-    int data_len = data.size();
+    Node snd_compact;
+    std::string snd_schema = "";
 
+    const void *snd_ptr = node.contiguous_data_ptr();
+    int         snd_data_size = node.total_bytes_compact();
+    
+    if(snd_ptr != NULL && 
+       node.is_compact() )
+    {
+        snd_schema = node.schema().to_json();
+    }
+    else
+    {
+        node.compact_to(snd_compact);
+        snd_schema = snd_compact.schema().to_json();
+        snd_ptr = snd_compact.data_ptr();
+    }
+    
+    int snd_schema_size = snd_schema.length() + 1;
 
-    int intArray[2] = { schema_len, data_len };
 
+    int mpi_error = MPI_Send(&snd_schema_size,
+                             1,
+                             MPI_INT,
+                             dest,
+                             tag,comm);
 
-    int mpi_error = MPI_Send(intArray, 2, MPI_INT, dest, tag, comm);
     CONDUIT_CHECK_MPI_ERROR(mpi_error);
 
-    mpi_error = MPI_Send(const_cast <char*> (schema.c_str()), schema_len, MPI_CHAR, dest, tag, comm);
+    mpi_error = MPI_Send(const_cast <char*> (snd_schema.c_str()),
+                         snd_schema_size,
+                         MPI_CHAR,
+                         dest,
+                         tag,
+                         comm);
+
     CONDUIT_CHECK_MPI_ERROR(mpi_error);
     
-    mpi_error = MPI_Send((char*)&data[0], data_len, MPI_CHAR, dest, tag, comm);
+    mpi_error = MPI_Send(const_cast<void*>(snd_ptr),
+                         snd_data_size,
+                         MPI_BYTE,
+                         dest,
+                         tag,
+                         comm);
+
     CONDUIT_CHECK_MPI_ERROR(mpi_error);
 
     return mpi_error;
 }
 
+
 //---------------------------------------------------------------------------//
 int
-recv(Node &node, int src, int tag, MPI_Comm comm)
+recv_using_schema(Node &node, int src, int tag, MPI_Comm comm)
 {  
-    int rcv_counts[2];
+    int rcv_schema_size = 0;
     MPI_Status status;
 
-    int mpi_error = MPI_Recv(rcv_counts, 2, MPI_INT, src, tag, comm, &status);
+    // TODO: use MPI_Probe?
+
+    int mpi_error = MPI_Recv(&rcv_schema_size,
+                             1,
+                             MPI_INT,
+                             src,
+                             tag,
+                             comm,
+                             &status);
+                             
+    CONDUIT_CHECK_MPI_ERROR(mpi_error);
+
+    Node rcv_buffers;
```

Comment:
I see your point -- schema payloads will always be char *s.  I will change this case. 

My main reason for doing this was to prevent any leaks. We will still need an optional Node buffer for the recv result (one which allocs for the proper type) so we definitely need to use a node there. 



For the more complex cases (ex: gather) I do feel using a node to manage the temp allocs and pointers needed makes things much nicer.